wsf_scraping/spiders/nice_spider.py

```diff
@@ -0,0 +1,227 @@
+                headers=headers,
+                callback=self.parse,
+            )
+
+    def parse(self, response):
+        """ Parse the guidance listing page. Request must be done in AJAX,
+        else the NICE website only send the default number of results (10).
+
+        @ajax
+        @url https://www.nice.org.uk/guidance/published/ajax?iDisplayLength=10
+        @returns items 0 0
+        @returns requests 10 30
+        """
+
+        # Grab the link to the detailed article, its evidences and history
+        try:
+            articles = json.loads(response._body)
+            doc_links = []
+            evidence_links = []
+            history_links = []
+
+            for doc in articles['aaData']:
+                doc_link = html.fromstring(doc['ProductTitle']).get('href')
+
+                doc_links.append('https://www.nice.org.uk%s' % doc_link)
+
+                if self.settings['NICE_GET_EVIDENCES']:
+                    evidence_links.append(
+                        'https://www.nice.org.uk%s/evidence' % doc_link
+                    )
+                if self.settings['NICE_GET_HISTORY']:
+                    history_links.append(
+                        'https://www.nice.org.uk%s/history' % doc_link
+                    )
+
+            for url in doc_links:
+                yield scrapy.Request(
+                    url=url,
+                    callback=self.parse_article
+                )
+
+            for url in evidence_links:
+                yield scrapy.Request(
+                    url=url,
+                    callback=self.parse_evidences
+                )
+
+            for url in history_links:
+                yield scrapy.Request(
+                    url=url,
+                    callback=self.parse_histories
+                )
+
+        except json.decoder.JSONDecodeError:
+            logging.info('Response was not json serialisable')
+
+    def parse_evidences(self, response):
+        """ Scrape the guidance evidencies. Then, redirect to the PDF pages.
+
+        @url https://www.nice.org.uk/guidance/ta494/evidence
+        @returns requests 1 1
+        @returns items 0 0
+        """
+
+        data_dict = {
+            'title': response.css('h1::text').extract_first(),
+            'year': response.css(
+                '.published-date time::attr(datetime)'
+            ).extract_first()[:4]
+        }
+        # Scrap all the pdf on the page, passing scrapped metadata
+        for href in response.css('.track-link::attr(href)').extract():
+            yield Request(
+                url=response.urljoin(href),
+                callback=self.save_pdf,
+                meta={'data_dict': data_dict}
+            )
+
+    def parse_histories(self, response):
+        """ Scrape the guidance history. Then redirect to the PDF pages.
+
+        @url https://www.nice.org.uk/guidance/ta494/history
+        @returns requests 24 24
+        @returns items 0 0
+        """
+
+        data_dict = {
```

Comment:
this code is being repeated, do you think it might be good to add it in a separate function?

Reply:
Hum, the first purpose was to parse evidences and history in different functions, for I used a different selector to retrieve the pdfs, but now I'm using the same one... Will definitely refactor!

Reply:
👍